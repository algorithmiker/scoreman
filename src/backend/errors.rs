use std::{
    cmp::{max, min},
    fmt::{self, Display},
    ops::RangeInclusive,
};

use yansi::Paint;

use crate::collect_parse_error;

#[derive(Clone, Debug)]
pub enum ErrorLocation {
    NoLocation,
    LineOnly(usize),
    LineAndMeasure(usize, usize),
    LineAndCharIdx(usize, usize),
}
/// Produced [diagnostics.len()] diagnostics and one error.
/// Diagnostics:
///  - [location]
///    [severity] [message]
///
/// [short]
/// [location]
/// [long]
///
#[derive(Debug)]
pub struct BackendError<'a> {
    pub main_location: ErrorLocation,
    pub relevant_lines: RangeInclusive<usize>,
    pub kind: BackendErrorKind<'a>,
    pub diagnostics: Vec<Diagnostic>,
}

#[derive(Debug)]
pub enum BackendErrorKind<'a> {
    IOError(std::io::Error),
    FmtError(std::fmt::Error),
    EmptyScore,
    /// string name and fret
    NoSuchFret(char, u16),
    TickMismatch(char, char, usize, usize),
    /// string name where tick is multichar, string name here, tick idx, and the found invalid fret
    BadMulticharTick {
        /// string and fret
        multichar: (char, u16),
        /// string and fret
        invalid: (char, u16),
        tick_idx: usize,
    },
    InvalidCommentSyntax(String),
    InvalidPartlineSyntax(String),
    ParseError(nom::Err<nom::error::VerboseError<&'a str>>),
}

impl<'a> BackendErrorKind<'a> {
    pub fn desc(&self) -> (String, String) {
        match self {
            BackendErrorKind::IOError(x) => {
                ("Cannot write to file".into(), format!("IO Error:\n{x}"))
            }

            BackendErrorKind::FmtError(x) => {
                ("Cannot write to internal buffer".into(), format!("Format error:\n{x}"))
            }
            BackendErrorKind::EmptyScore => ("Empty score".into(), String::new()),
            BackendErrorKind::NoSuchFret(string_name, fret) => (
                "No such fret".into(),
                format!("Failed to get note for fret {fret} on string {string_name}"),
            ),

            BackendErrorKind::TickMismatch(string_before, string_after,ticks_before, ticks_after) => ("Tick mismatch".into(),
format!("The muxml2 backend relies on the fact that there are the same number of ticks (frets/rests) on every line (string) of a measure in the tab. This is not true for this tab.
The measure has {ticks_before} ticks on string {string_before} and {ticks_after} ticks on string {string_after}.

Tip: If you get a lot of errors like this, consider using the muxml1 backend.")
            ),
            BackendErrorKind::BadMulticharTick { multichar : (multichar_string,multichar_fret), invalid: (invalid_string,invalid_fret), tick_idx } =>
            (
                "Invalid multichar tick".into(), 
                format!(
"Tick {} has a multi-char fret ({multichar_fret}) on string {multichar_string}, but on the same tick there is an invalid fret {invalid_fret} on string {invalid_string}", tick_idx+1)
            ),
            BackendErrorKind::InvalidCommentSyntax(rem) => ("Invalid comment syntax".into(), format!("Got remaining content: `{rem}`")),
            BackendErrorKind::InvalidPartlineSyntax(rem) => ("Invalid partline syntax".into(), format!("Got remaining content: `{rem}`")),
            BackendErrorKind::ParseError(x) => ("Invalid syntax".into(), collect_parse_error(x)),

        }
    }
}
#[derive(Debug, Clone)]
pub enum DiagnosticSeverity {
    Info,
    Warning,
}
impl Display for DiagnosticSeverity {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{}",
            match self {
                DiagnosticSeverity::Info => "Info".blue(),
                DiagnosticSeverity::Warning => "Warning".yellow(),
            }
        )
    }
}
#[derive(Debug, Clone)]
pub struct Diagnostic {
    pub location: ErrorLocation,
    pub message: String,
    pub severity: DiagnosticSeverity,
}

impl Diagnostic {
    pub fn info(location: ErrorLocation, message: String) -> Self {
        Self {
            location,
            message,
            severity: DiagnosticSeverity::Info,
        }
    }
    pub fn warn(location: ErrorLocation, message: String) -> Self {
        Self {
            location,
            message,
            severity: DiagnosticSeverity::Warning,
        }
    }
}

impl<'a> BackendError<'a> {
    pub fn from_io_error(x: std::io::Error, diagnostics: Vec<Diagnostic>) -> Self {
        BackendError {
            main_location: ErrorLocation::NoLocation,
            kind: BackendErrorKind::IOError(x),
            relevant_lines: 0..=0,
            diagnostics,
        }
    }
    pub fn from_fmt_error(x: std::fmt::Error, diagnostics: Vec<Diagnostic>) -> Self {
        BackendError {
            main_location: ErrorLocation::NoLocation,
            //short: "Cannot write to internal buffer".to_string(),
            //long: format!("Format error:\n{x}"),
            diagnostics,
            relevant_lines: 0..=0,
            kind: BackendErrorKind::FmtError(x),
        }
    }

    pub fn empty_score_err(diagnostics: Vec<Diagnostic>) -> Self {
        BackendError {
            main_location: ErrorLocation::NoLocation,
            diagnostics,
            kind: BackendErrorKind::EmptyScore,
            relevant_lines: 0..=0,
        }
    }

    pub fn no_such_fret(
        location_a: usize,
        location_b: usize,
        string_name: char,
        fret: u16,
        diagnostics: Vec<Diagnostic>,
    ) -> Self {
        Self {
            main_location: ErrorLocation::LineAndMeasure(location_a, location_b),
            diagnostics,
            kind: BackendErrorKind::NoSuchFret(string_name, fret),
            relevant_lines: location_a..=location_a,
        }
    }
}

impl ErrorLocation {
    pub fn get_line_idx(&self) -> Option<usize> {
        match self {
            ErrorLocation::NoLocation => None,
            ErrorLocation::LineOnly(x) => Some(*x),
            ErrorLocation::LineAndMeasure(x, _) => Some(*x),
            ErrorLocation::LineAndCharIdx(x, _) => Some(*x),
        }
    }
    pub fn write_location_explainer(&self, f: &mut impl std::fmt::Write) {
        match self {
            ErrorLocation::NoLocation => (),
            ErrorLocation::LineOnly(line_idx) => {
                let line_num = line_idx + 1;
                writeln!(f, "{} in line {line_num}:", "Where:".bold(),).unwrap();
            }
            ErrorLocation::LineAndMeasure(line_idx, measure_idx) => {
                let (line_num, measure_num) = (line_idx + 1, measure_idx + 1);
                writeln!(
                    f,
                    "{} Measure {measure_num} in line {line_num}:",
                    "Where:".bold()
                )
                .unwrap();
            }
            ErrorLocation::LineAndCharIdx(line_idx, char_idx) => writeln!(
                f,
                "{} line {} char {}",
                "Where:".bold(),
                line_idx + 1,
                char_idx + 1,
            )
            .unwrap(),
        }
    }
}

pub const ERROR_CONTEXT: usize = 3;

/// When reporting an error with `relevant_lines`, we want to show some context of +=, [ERROR_CONTEXT]
/// lines that are not neccessarily relevant
/// This is not always possible because of line bounds (cannot show 3 lines before line 0)
/// This function handles that.
pub fn extend_error_range(range: &RangeInclusive<usize>, line_cnt: usize) -> RangeInclusive<usize> {
    let start = max(*range.start(), ERROR_CONTEXT) - ERROR_CONTEXT;
    let end = min(line_cnt, range.end() + ERROR_CONTEXT);

    start..=end
}

pub fn get_digit_cnt(num: usize) -> u32 {
    num.checked_ilog10().unwrap_or(0) + 1
}
