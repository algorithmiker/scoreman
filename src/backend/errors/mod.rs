use std::{
    cmp::{max, min},
    ops::RangeInclusive,
};

pub mod backend_error;
pub mod backend_error_kind;
pub mod diagnostic;
pub mod diagnostic_kind;
pub mod error_location;

pub const ERROR_CONTEXT: usize = 3;

/// When reporting an error with `relevant_lines`, we want to show some context of +=, [ERROR_CONTEXT]
/// lines that are not neccessarily relevant
/// This is not always possible because of line bounds (cannot show 3 lines before line 0)
/// This function handles that.
pub fn extend_error_range(range: &RangeInclusive<usize>, line_cnt: usize) -> RangeInclusive<usize> {
    let start = max(*range.start(), ERROR_CONTEXT) - ERROR_CONTEXT;
    let end = min(line_cnt - 1, range.end() + ERROR_CONTEXT);

    start..=end
}
