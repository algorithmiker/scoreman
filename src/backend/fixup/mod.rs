use std::time::{Duration, Instant};

use clap::ValueEnum;

use super::BackendResult;
use crate::{
    backend::{
        errors::{
            backend_error::BackendError, backend_error_kind::BackendErrorKind,
            diagnostic::Diagnostic, diagnostic_kind::DiagnosticKind, error_location::ErrorLocation,
        },
        Backend,
    },
    parser::parser::parse,
    time, traceln,
};

pub struct FixupBackend();
#[derive(ValueEnum, Clone)]
pub enum FixupDumpOptions {
    TickStream,
    PrettyTracks,
}
#[derive(Clone)]
pub struct FixupBackendSettings {
    pub dump: Option<FixupDumpOptions>,
}

struct LocationTracker {
    pub data: [ErrorLocation; 5],
    pub push_cnt: u32,
}

impl LocationTracker {
    pub fn new() -> Self {
        Self { data: [const { ErrorLocation::NoLocation }; 5], push_cnt: 0 }
    }
    pub fn add(&mut self, l: ErrorLocation) {
        traceln!("location_tracker:: before add: {:?}", self.data);
        // shift left
        for i in 0..4 {
            self.data[i] = self.data[i + 1].clone();
        }
        self.data[4] = l;
        traceln!("location_tracker:: after add: {:?}", self.data);
        self.push_cnt += 1;
    }
    pub fn is_same(&self) -> bool {
        self.push_cnt > 4 && self.data.windows(2).all(|a| a[0] == a[1])
    }
}
impl Backend for FixupBackend {
    type BackendSettings = FixupBackendSettings;

    fn process<Out: std::io::Write>(
        parser_input: &[String], out: &mut Out, settings: Self::BackendSettings,
    ) -> BackendResult {
        if let Some(dump) = settings.dump {
            let (parse_time, parsed) = time(|| parse(parser_input));
            let mut r = BackendResult::new(vec![], None, Some(parse_time), None);
            match dump {
                FixupDumpOptions::TickStream => writeln!(out, "{:?}", parsed.tick_stream).unwrap(),
                FixupDumpOptions::PrettyTracks => {
                    writeln!(out, "{}", parsed.dump_tracks()).unwrap()
                }
            }
            r.err = parsed.error;
            return r;
        }
        let mut diagnostics = vec![];
        // TODO: figure out a way not to clone these
        let mut parser_input = parser_input.to_owned();
        let mut parse_time = Duration::from_secs(0);
        let fixup_start = Instant::now();
        let mut location_tracker = LocationTracker::new();
        loop {
            let parse_start = Instant::now();
            let parsed = parse(&parser_input);
            parse_time = parse_start.elapsed();
            match &parsed.error {
                None => break,
                Some(err) => {
                    location_tracker.add(err.main_location.clone());
                    if location_tracker.is_same() {
                        let lines =
                            err.main_location.get_line_idx().map(|x| x..=x).unwrap_or(0..=0);
                        return BackendResult::new(
                            diagnostics,
                            Some(BackendError::fixup_failed(err.main_location.clone(), lines)),
                            Some(parse_time),
                            Some(fixup_start.elapsed()),
                        );
                    }
                    match err.kind {
                        BackendErrorKind::IOError(_) | BackendErrorKind::FmtError(_) => {
                            return BackendResult::new(
                                diagnostics,
                                parsed.error,
                                Some(parse_time),
                                None,
                            );
                        }
                        BackendErrorKind::BendOnInvalid => {} // todo: bendOnInvalid fixup: remove the bend
                        BackendErrorKind::InvalidStringName => {}
                        BackendErrorKind::EmptyScore => {}
                        BackendErrorKind::BothSlotsMultiChar => {} // todo: fix BothSlotsMultichar errors
                        BackendErrorKind::FretTooLarge => {} // todo: fix FretTooLarge errors (add
                        // space between)
                        BackendErrorKind::MultiBothSlotsFilled => {
                            let Some((line_idx, char_idx)) = err
                                .main_location
                                .get_line_idx()
                                .zip(err.main_location.get_char_idx())
                            else {
                                continue;
                            };
                            // we just replace both with a rest to be sure
                            // todo: use a better strategy by checking the actual ticks
                            parser_input[line_idx].replace_range(char_idx..char_idx + 1, "-");
                            parser_input[line_idx].replace_range(char_idx + 1..char_idx + 2, "-");
                            diagnostics.push(Diagnostic::info(
                                err.main_location.clone(),
                                DiagnosticKind::FormatReplacedInvalid,
                            ));
                        }
                        BackendErrorKind::NoClosingBarline => {
                            let l_idx = err.main_location.get_line_idx().unwrap();
                            let line = &mut parser_input[l_idx];
                            line.truncate(line.trim_end().len());
                            line.push('|');
                            diagnostics.push(Diagnostic::info(
                                err.main_location.clone(),
                                DiagnosticKind::FormatAddedBarline,
                            ))
                        }
                        BackendErrorKind::FixupFailed => unreachable!(),
                        BackendErrorKind::Parse3InvalidCharacter(_) => {
                            let Some((line_idx, char_idx)) = err
                                .main_location
                                .get_line_idx()
                                .zip(err.main_location.get_char_idx())
                            else {
                                continue;
                            };
                            parser_input[line_idx].replace_range(char_idx..char_idx + 1, "-");
                            diagnostics.push(Diagnostic::info(
                                err.main_location.clone(),
                                DiagnosticKind::FormatReplacedInvalid,
                            ))
                        }
                    };
                }
            }
        }
        let gen_time = fixup_start.elapsed();
        let maybe_io_err =
            out.write_all(parser_input.join("\n").as_ref()).map_err(BackendError::from).err();
        BackendResult::new(diagnostics, maybe_io_err, Some(parse_time), Some(gen_time))
    }
}
