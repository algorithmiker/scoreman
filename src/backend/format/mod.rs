use super::{Backend, BackendError, Diagnostic};
use crate::parser::Section;

pub struct FormatBackend();

impl Backend for FormatBackend {
    type BackendSettings = ();

    fn process<Out: std::io::Write>(
        score: crate::parser::Score,
        out: &mut Out,
        _settings: Self::BackendSettings,
    ) -> Result<Vec<Diagnostic>, BackendError> {
        let diagnostics = vec![];
        let mut formatted = String::new();
        let mut measure_cnt = 0;
        for section in score.0 {
            match section {
                Section::Part { part, .. } => {
                    let measures_in_part = part[0].staffs.len();
                    for measure_idx in 0..measures_in_part {
                        formatted += &format!("// SYS: Measure {}\n", measure_cnt + 1);
                        for line in &part {
                            formatted.push(line.string_name);
                            formatted.push('|');
                            formatted += &line.staffs[measure_idx].print_pretty_string();
                            formatted.push('|');
                            formatted.push('\n');
                        }
                        formatted.push('\n');
                        measure_cnt += 1;
                    }
                }
                Section::Comment(x) => {
                    // SYS-comments were generated during a previous format run, so don't include
                    // them
                    if !x.trim_start().starts_with("SYS:") {
                        formatted += "//";
                        formatted += &x;
                        formatted.push('\n');
                    }
                }
            }
        }

        if let Err(x) = out.write_all(formatted.as_bytes()) {
            return Err(BackendError::from_io_error(x, diagnostics));
        }

        Ok(diagnostics)
    }
}
