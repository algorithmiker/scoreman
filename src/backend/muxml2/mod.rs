pub mod fretboard;
mod muxml2_formatters;
#[cfg(test)]
mod muxml2_tests;
pub mod settings;

use muxml2_formatters::{
    write_muxml2_measure, write_muxml2_note, write_muxml2_rest, MUXML2_DOCUMENT_END,
    MUXML_INCOMPLETE_DOC_PRELUDE,
};

use crate::{
    backend::errors::error_location::ErrorLocation,
    parser::{
        Score,
        TabElement::{self, Fret, Rest},
    },
    raw_tracks::RawTracks,
};

use self::fretboard::get_fretboard_note;

use super::{
    errors::{
        backend_error::BackendError, backend_error_kind::BackendErrorKind, diagnostic::Diagnostic,
    },
    Backend,
};

pub struct Muxml2Backend();
impl Backend for Muxml2Backend {
    type BackendSettings = settings::Settings;

    fn process<Out: std::io::Write>(
        score: Score,
        out: &mut Out,
        settings: Self::BackendSettings,
    ) -> Result<Vec<Diagnostic>, BackendError> {
        let mut diagnostics = vec![];
        let (raw_tracks, tick_cnt) = score.gen_raw_tracks()?;
        let (xml_out, mut inner_diagnostics) =
            raw_tracks_to_muxml2(raw_tracks, settings, tick_cnt)?;
        diagnostics.append(&mut inner_diagnostics);
        out.write_all(xml_out.as_bytes())
            .map_err(|x| BackendError::from_io_error(x, diagnostics.clone()))?;

        Ok(diagnostics)
    }
}

#[derive(Debug)]
pub enum Muxml2TabElement {
    Rest(usize),
    Notes(Vec<MuxmlNote>),
    /// used in optimizing, should generate no code for this type
    Invalid,
}

impl Muxml2TabElement {
    fn write_muxml<A: std::fmt::Write>(&self, buf: &mut A) -> std::fmt::Result {
        match self {
            Muxml2TabElement::Rest(mut x) => {
                while x != 0 {
                    if x >= 8 {
                        write_muxml2_rest(buf, "whole", 8)?;
                        x -= 8;
                    } else if x >= 4 {
                        write_muxml2_rest(buf, "half", 4)?;
                        x -= 4;
                    } else if x >= 2 {
                        write_muxml2_rest(buf, "quarter", 2)?;
                        x -= 2;
                    } else {
                        debug_assert_eq!(x, 1);
                        write_muxml2_rest(buf, "eighth", 1)?;
                        x -= 1;
                    }
                }
                Ok(())
            }
            Muxml2TabElement::Notes(notes) => {
                for (i, note) in notes.iter().enumerate() {
                    note.write_muxml(buf, "eighth", i != 0)?;
                }
                Ok(())
            }
            Muxml2TabElement::Invalid => Ok(()),
        }
    }
}

fn raw_tracks_to_muxml2<'a>(
    raw_tracks: RawTracks,
    settings: <Muxml2Backend as Backend>::BackendSettings,
    tick_cnt: usize,
) -> Result<(String, Vec<Diagnostic>), BackendError<'a>> {
    // the muxml2 backend assumes
    // 1. that there are the same number of measures for every string (which should be true)
    // 2. that there are the same number of elements in the same measure for each string (also
    //    generally true)
    let diagnostics = vec![];
    let number_of_measures = raw_tracks.1[0].len();
    let mut document = String::from(MUXML_INCOMPLETE_DOC_PRELUDE);
    // this looks like a good setting for -nmt based on trial and error
    document.reserve(tick_cnt * 10);
    //println!("Reserved capacity: {}", document.capacity());
    for measure_idx in 0..number_of_measures {
        let ticks_in_measure = raw_tracks.1[0][measure_idx].content.len();

        // Length of actual content in measure. `remove_space_between_notes` will reduce this for
        // example
        let mut measure_content_len = ticks_in_measure;
        let mut measure_processed: Vec<Muxml2TabElement> = vec![];
        for tick in 0..ticks_in_measure {
            let mut notes_in_tick = vec![];
            for string_idx in 0..6 {
                let measure = &raw_tracks.1[string_idx][measure_idx];
                let raw_tick = match measure.content.get(tick) {
                    Some(x) => x,
                    _ => {
                        return _tick_mismatch_err(
                            raw_tracks,
                            string_idx,
                            measure_idx,
                            diagnostics,
                        );
                    }
                };
                let loc = (measure.parent_line, measure.index_on_parent_line);
                match raw_tick.element {
                    Fret(fret) => {
                        let x =
                            get_fretboard_note(raw_tracks.0[string_idx], fret, loc, &diagnostics)?;
                        notes_in_tick.push(x);
                    }
                    TabElement::DeadNote => {
                        let mut x =
                            get_fretboard_note(raw_tracks.0[string_idx], 0, loc, &diagnostics)?;
                        x.dead = true;
                        notes_in_tick.push(x);
                    }
                    Rest => continue,
                }
            }
            // if there were no notes inserted in this tick, add a rest
            measure_processed.push(if notes_in_tick.is_empty() {
                Muxml2TabElement::Rest(1)
            } else {
                Muxml2TabElement::Notes(notes_in_tick)
            })
        }
        //println!("[D]: Measure before opt: {measure_processed:?}");

        // remove rest between notes if wanted
        if settings.remove_rest_between_notes {
            let mut i = 0;
            while i < measure_processed.len() {
                use Muxml2TabElement::*;
                match (
                    measure_processed.get(i),
                    measure_processed.get(i + 1),
                    measure_processed.get(i + 2),
                ) {
                    (Some(Notes(_)), Some(Rest(1)), Some(Notes(_))) => {
                        measure_processed[i + 1] = Muxml2TabElement::Invalid;
                        i += 3;
                        measure_content_len -= 1;
                    }
                    (Some(Rest(1)), Some(Notes(_)), Some(Rest(1))) => {
                        measure_processed[i] = Muxml2TabElement::Invalid;
                        measure_processed[i + 2] = Muxml2TabElement::Invalid;
                        i += 3;
                        measure_content_len -= 2;
                    }
                    _ => {
                        i += 1;
                    }
                }
            }
        }

        // merge rests in measure
        for mut i in 0..measure_processed.len() {
            match measure_processed[i] {
                Muxml2TabElement::Rest(x) => {
                    debug_assert_eq!(x, 1, "Expect Muxml2TabElement::Rest(1) in unprocessed AST, got Muxml2TabElement::Rest({x})");
                    let original_i = i;
                    while i < measure_processed.len()
                        && matches!(measure_processed[i], Muxml2TabElement::Rest(1))
                    {
                        measure_processed[i] = Muxml2TabElement::Invalid;
                        i += 1;
                    }
                    measure_processed[original_i] = Muxml2TabElement::Rest(i - original_i);
                }
                Muxml2TabElement::Notes(..) | Muxml2TabElement::Invalid => continue,
            }
        }

        if settings.trim_measure {
            trim_measure(
                &mut measure_processed,
                &mut measure_content_len,
                Direction::Forward,
            );
            trim_measure(
                &mut measure_processed,
                &mut measure_content_len,
                Direction::Backward,
            );
        }

        // write the final contents into a buffer
        let mut measure_xml = String::new();
        for proc_elem in measure_processed {
            if let Err(x) = proc_elem.write_muxml(&mut measure_xml) {
                return Err(BackendError::from_fmt_error(x, diagnostics));
            }
        }

        // Try to simplify e.g 8/8 to 4/4
        let (mut measure_enumerator, mut measure_denominator) = (measure_content_len, 8);
        if settings.simplify_time_signature && measure_content_len % 2 == 0 {
            measure_enumerator /= 2;
            measure_denominator /= 2;
        }
        write_muxml2_measure(
            &mut document,
            measure_idx,
            measure_enumerator,
            measure_denominator,
            &measure_xml,
        )
        .unwrap();
    }

    document += MUXML2_DOCUMENT_END;
    //println!("Actual len: {}", document.len());
    Ok((document, diagnostics))
}

enum Direction {
    Forward,
    Backward,
}

fn trim_measure(measure: &mut [Muxml2TabElement], content_len: &mut usize, direction: Direction) {
    let (mut i, last) = match direction {
        Direction::Forward => (0, measure.len() - 1),
        Direction::Backward => (measure.len() - 1, 0),
    };
    loop {
        match &measure[i] {
            Muxml2TabElement::Rest(rest_len) => {
                *content_len -= rest_len;
                measure[i] = Muxml2TabElement::Invalid;
                break;
            }
            Muxml2TabElement::Notes(_) => break,
            Muxml2TabElement::Invalid => {
                if i == last {
                    break;
                }
                match direction {
                    Direction::Forward => i += 1,
                    Direction::Backward => i -= 1,
                }
            }
        }
    }
}

#[derive(Clone, Debug)]
pub struct MuxmlNote {
    pub step: char,  // 1 byte
    pub octave: u8,  // 1 byte
    pub sharp: bool, // 1 byte
    pub dead: bool,  // 1 bytes
}

impl MuxmlNote {
    pub fn write_muxml(
        &self,
        buf: &mut impl std::fmt::Write,
        duration: &str,
        chord: bool,
    ) -> Result<(), std::fmt::Error> {
        write_muxml2_note(
            buf,
            self.step,
            self.octave,
            self.sharp,
            duration,
            chord,
            self.dead,
        )
    }
}

fn _tick_mismatch_err<T>(
    raw_tracks: RawTracks,
    string_idx: usize,
    measure_idx: usize,
    diagnostics: Vec<Diagnostic>,
) -> Result<T, BackendError<'static>> {
    let before_measure = &raw_tracks.1[string_idx - 1][measure_idx];
    let this_measure = &raw_tracks.1[string_idx][measure_idx];

    Err(BackendError {
        main_location: ErrorLocation::LineAndMeasure(
            this_measure.parent_line,
            this_measure.index_on_parent_line,
        ),
        diagnostics,
        relevant_lines: before_measure.parent_line..=this_measure.parent_line,
        kind: BackendErrorKind::TickMismatch(
            raw_tracks.0[string_idx - 1],
            raw_tracks.0[string_idx],
            before_measure.content.len(),
            this_measure.content.len(),
        ),
    })
}
