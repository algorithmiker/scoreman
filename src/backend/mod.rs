use errors::{backend_error::BackendError, diagnostic::Diagnostic};

use std::{fmt::Display, time::Duration};
pub mod errors;
pub mod fixup;
pub mod midi;
pub mod muxml;
pub mod muxml2;
pub struct BackendResult {
    pub diagnostics: Vec<Diagnostic>,
    pub err: Option<BackendError>,
    pub timing_parse: Option<Duration>,
    pub timing_gen: Option<Duration>,
}
impl BackendResult {
    pub fn new(
        diagnostics: Vec<Diagnostic>, err: Option<BackendError>, timing_parse: Option<Duration>,
        timing_gen: Option<Duration>,
    ) -> Self {
        Self { diagnostics, err, timing_parse, timing_gen }
    }
}
pub trait Backend {
    type BackendSettings;

    fn process<Out: std::io::Write>(
        input: &[String], out: &mut Out, settings: Self::BackendSettings,
    ) -> BackendResult;
}

/// Handles backend dispatch. Can be easily created from a string identifier
#[derive(Clone)]
pub enum BackendSelector {
    Midi(()),
    // Muxml(()),
    Muxml2(muxml2::settings::Settings),
    Fixup(fixup::FixupBackendSettings),
}

impl BackendSelector {
    // TODO: remove the lifetime here
    pub fn process<Out: std::io::Write>(self, input: &[String], out: &mut Out) -> BackendResult {
        match self {
            BackendSelector::Midi(settings) => midi::MidiBackend::process(input, out, settings),
            // BackendSelector::Muxml(settings) => muxml::MuxmlBackend::process(input, out, settings),
            BackendSelector::Muxml2(settings) => {
                muxml2::Muxml2Backend::process(input, out, settings)
            }
            BackendSelector::Fixup(settings) => fixup::FixupBackend::process(input, out, settings),
        }
    }
}

impl Display for BackendSelector {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}",
            match self {
                BackendSelector::Midi(_) => "midi",
                // BackendSelector::Muxml(_) => "muxml",
                BackendSelector::Muxml2(_) => "muxml",
                BackendSelector::Fixup(_) => "fixup",
            }
        )
    }
}
