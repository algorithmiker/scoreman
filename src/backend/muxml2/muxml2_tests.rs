use crate::backend::{
    muxml2::{settings::Settings, Muxml2Backend},
    Backend,
};

#[test]
fn test_muxml2() -> anyhow::Result<()> {
    let i1 = r#"
e|--------------|
B|-----------0-1|
G|-------1-2----|
D|--0-2-4-------|
A|-3------------|
E|--------------|
    "#;
    let mut out = vec![];
    let settings = Settings {
        remove_rest_between_notes: true,
        trim_measure: true,
        simplify_time_signature: true,
        bend_mode: crate::backend::muxml2::settings::Muxml2BendMode::EmulateBends,
    };
    Muxml2Backend::process(i1.lines(), &mut out, settings);
    insta::assert_snapshot!(String::from_utf8_lossy(&out));
    Ok(())
}
#[test]
fn test_muxml_bends() -> anyhow::Result<()> {
    let i1 = r#"
e|----|
B|----|
G|2b4-|
D|----|
A|----|
E|----|
    "#;
    let mut out = vec![];
    let settings = Settings {
        remove_rest_between_notes: true,
        trim_measure: true,
        simplify_time_signature: true,
        bend_mode: crate::backend::muxml2::settings::Muxml2BendMode::EmulateBends,
    };
    Muxml2Backend::process(i1.lines(), &mut out, settings);
    insta::assert_snapshot!(String::from_utf8_lossy(&out));
    Ok(())
}
