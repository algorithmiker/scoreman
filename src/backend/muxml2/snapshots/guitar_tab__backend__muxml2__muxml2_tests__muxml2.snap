---
source: src/backend/muxml2/muxml2_tests.rs
expression: "String::from_utf8_lossy(&out)"
---
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE score-partwise PUBLIC "-//Recordare//DTD MusicXML 4.0 Partwise//EN" "http://www.musicxml.org/dtds/partwise.dtd">
<score-partwise version="4.0">
  <identification>
    <encoding>
      <software>guitar_tab</software>
      <supports element="accidental" type="yes"/>
      <supports element="beam" type="yes"/>
      <supports element="print" attribute="new-page" type="no"/>
      <supports element="print" attribute="new-system" type="no"/>
      <supports element="stem" type="yes"/>
    </encoding>
  </identification>
  <part-list>
    <score-part id="P1">
      <part-name>Guitar1</part-name>
    </score-part>
  </part-list>
  <part id="P1">
<measure number="0">
<attributes>
<divisions>2</divisions>
<key><fifths>0</fifths></key>
<time><beats>9</beats><beat-type>8</beat-type></time>
<clef><sign>G</sign><line>2</line></clef>
</attributes>
<note>
<pitch><step>C</step>
<octave>4</octave>
</pitch>
<duration>1</duration>
<type>eighth</type></note>
<note>
<pitch><step>D</step>
<octave>4</octave>
</pitch>
<duration>1</duration>
<type>eighth</type></note>
<note>
<pitch><step>E</step>
<octave>4</octave>
</pitch>
<duration>1</duration>
<type>eighth</type></note>
<note>
<rest measure="no"/>
<duration>1</duration>
<voice>1</voice>
<type>eighth</type>
</note>
<note>
<pitch><step>F</step>
<alter>1</alter>
<octave>4</octave>
</pitch>
<duration>1</duration>
<type>eighth</type><accidental>sharp</accidental>
</note>
<note>
<pitch><step>G</step>
<alter>1</alter>
<octave>4</octave>
</pitch>
<duration>1</duration>
<type>eighth</type><accidental>sharp</accidental>
</note>
<note>
<pitch><step>A</step>
<octave>4</octave>
</pitch>
<duration>1</duration>
<type>eighth</type></note>
<note>
<pitch><step>B</step>
<octave>4</octave>
</pitch>
<duration>1</duration>
<type>eighth</type></note>
<note>
<pitch><step>C</step>
<octave>5</octave>
</pitch>
<duration>1</duration>
<type>eighth</type></note>
</measure>
</part>
</score-partwise>
